Handle different mappings for _kuzzle_info between indices

The default mapping for `_kuzzle_info` has changed between kuzzle v1.2.6 and 1.2.7

When working with some indices created before version 1.2.7 and when upgrading Kuzzle, Kuzzle complains at startup that the new mapping cannot be applied.

We need Kuzzle to allow different mapping for the field between different indicies as this has no functional impact (only the index size in Elasticsearch for the new mapping are lighter).

## Expected Behavior
Be able to handle indices with different historical compatible versions of the mapping for _kuzzle_info

## Current Behavior
`kuzzle              | 0|KuzzleSe | [x] [ERROR] BadRequestError: Can not change type of field "_kuzzle_info.updater" from "text" to "keyword"`

## Possible Solution
Inject the mapping for the field only if it does not exist yet.

--- a/lib/api/core/indexCache.js
+++ b/lib/api/core/indexCache.js
@@ -20,7 +20,9 @@
  */
 'use strict';
 
-const Bluebird = require('bluebird');
+const
+  _ = require('lodash'),
+  Bluebird = require('bluebird');
 
 /**
  * Index/collection cache management
@@ -29,6 +31,8 @@ class IndexCache {
 
   constructor (kuzzle) {
     this.indexes = {};
+    this.defaultMappings = {};
+
     this.kuzzle = kuzzle;
     this.commonMapping = this.kuzzle.config.services.db.commonMapping;
   }
@@ -36,29 +40,58 @@ class IndexCache {
   initInternal (internalEngine) {
     return internalEngine.getMapping()
       .then(mapping => {
-        Object.keys(mapping).forEach(index => {
+        for (const index of Object.keys(mapping)) {
           this.indexes[index] = Object.keys(mapping[index].mappings);
-        });
+        }
       });
   }
 
   init () {
-    return this.kuzzle.internalEngine.listIndexes()
+    const engine = this.kuzzle.internalEngine;
+    const kuzzleInfoFieldsRegex = new RegExp(`^_kuzzle_info\\.(${Object.keys(this.commonMapping._kuzzle_info.properties).join('|')})$`);
+
+    return engine.listIndexes()
       .then(indexes => {
         const promises = [];
 
         for (const index of indexes) {
-          this.indexes[index] = [];
+          if (index === this.kuzzle.internalEngine.index) {
+            continue;
+          }
 
-          promises.push(this.kuzzle.internalEngine.listCollections(index)
+          this.indexes[index] = [];
+          this.defaultMappings[index] = _.cloneDeep(this.commonMapping);
+
+          promises.push(engine.getFieldMapping({
+            index,
+            fields: '_kuzzle_info.*'
+          })
+            .then(kuzInfoMappings => {
+              if (kuzInfoMappings[index] && kuzInfoMappings[index].mappings) {
+                for (const collection of Object.keys(kuzInfoMappings[index].mappings)) {
+                  const collectionInfo = kuzInfoMappings[index].mappings[collection];
+
+                  for (const dotField of Object.keys(collectionInfo).filter(f => kuzzleInfoFieldsRegex.test(f))) {
+                    Object.assign(this.defaultMappings[index]._kuzzle_info.properties, collectionInfo[dotField].mapping);
+                  }
+                  // do not break to catch cases where not all fields are defined
+                }
+              }
+
+              return engine.listCollections(index);
+            })
             .then(collections => {
+              const updateMappingPromises = [];
+
               this.indexes[index] = collections;
 
-              const mapping = {
-                properties: this.commonMapping
-              };
+              for (const collection of collections) {
+                updateMappingPromises.push(engine.updateMapping(collection, {
+                  properties: this.defaultMappings[index]
+                }, index));
+              }
 
-              return Bluebird.map(collections, collection => this.kuzzle.internalEngine.updateMapping(collection, mapping, index));
+              return Bluebird.all(updateMappingPromises);
             })
           );
         }

--- a/lib/services/elasticsearch.js
+++ b/lib/services/elasticsearch.js
@@ -574,10 +574,11 @@ class ElasticSearch extends Service {
 
     const requestBody = request.input.body || {};
     const collectionMapping = requestBody.properties || {};
+    const defaultMapping = this.kuzzle.indexCache.defaultMappings[esRequest.index] || this.config.commonMapping;
 
     esRequest.body = {
       [esRequest.type]: {
-        properties: Object.assign({}, collectionMapping, this.config.commonMapping)
+        properties: Object.assign({}, collectionMapping, defaultMapping)
       }
     };
 
@@ -730,14 +731,28 @@ class ElasticSearch extends Service {
    */
   updateMapping(request) {
     const esRequest = initESRequest(request);
+    const defaultMapping = this.kuzzle.indexCache.defaultMappings[esRequest.index] || this.config.commonMapping;
 
     esRequest.body = request.input.body;
     if (!esRequest.body.properties) {
       esRequest.body.properties = {};
     }
-    esRequest.body.properties = Object.assign({}, this.config.commonMapping, esRequest.body.properties);
+    esRequest.body.properties = Object.assign({}, defaultMapping, esRequest.body.properties);
 
     return this.client.indices.putMapping(esRequest)
+      .catch(error => Bluebird.reject(this.esWrapper.formatESError(error)));
+  }
+
+  /**
+   * Retrieve a mapping for a given field or collection of fields
+   *
+   * @param {Request} request
+   * @returns {Promise}
+   */
+  getFieldMapping(request) {
+    const esRequest = initESRequest(request, ['fields']);
+
+    return this.client.indices.getFieldMapping(esRequest)
       .catch(error => Bluebird.reject(this.esWrapper.formatESError(error)));
   }
 

--- a/lib/services/internalEngine/index.js
+++ b/lib/services/internalEngine/index.js
@@ -446,6 +446,17 @@ class InternalEngine {
   }
 
   /**
+   * Retrieves the mapping of one or several given fields
+   *
+   * @param data
+   * @returns {Promise}
+   */
+  getFieldMapping (data) {
+    return this.client.indices.getFieldMapping(data)
+      .catch(error => this.esWrapper.formatESError(error));
+  }
+
+  /**
    * Retrieve mapping definiton of index or index/collection
    *
    * @returns {Promise}

--- a/test/api/core/indexCache.test.js
+++ b/test/api/core/indexCache.test.js
@@ -1,24 +1,21 @@
-var
+let
   should = require('should'),
-  Promise = require('bluebird'),
   sinon = require('sinon'),
   sandbox = sinon.sandbox.create(),
   KuzzleMock = require('../../mocks/kuzzle.mock'),
   IndexCache = require.main.require('lib/api/core/indexCache');
 
 describe('Test: core/indexCache', () => {
-  var
+  let
     listIndexesStub,
     listCollectionsStub,
     getMappingStub,
     indexCache,
     kuzzle;
 
-  before(() => {
-    kuzzle = new KuzzleMock();
-  });
-
   beforeEach(() => {
+    kuzzle = new KuzzleMock();
+
     const internalMapping = {
       foo: {
         mappings: {
@@ -29,9 +26,10 @@ describe('Test: core/indexCache', () => {
       }
     };
 
-    listIndexesStub = kuzzle.internalEngine.listIndexes.returns(Promise.resolve(['foo']));
-    listCollectionsStub = kuzzle.internalEngine.listCollections.returns(Promise.resolve(['bar', 'baz', 'qux']));
-    getMappingStub = kuzzle.internalEngine.getMapping.returns(Promise.resolve(internalMapping));
+    listIndexesStub = kuzzle.internalEngine.listIndexes.resolves(['foo']);
+    listCollectionsStub = kuzzle.internalEngine.listCollections.resolves(['bar', 'baz', 'qux']);
+    getMappingStub = kuzzle.internalEngine.getMapping.resolves(internalMapping);
+    kuzzle.internalEngine.getFieldMapping.resolves({});
     indexCache = new IndexCache(kuzzle);
   });
 
@@ -41,6 +39,8 @@ describe('Test: core/indexCache', () => {
 
   describe('#init', () => {
     it('should initialize the index cache properly', () => {
+      kuzzle.internalEngine.getFieldMapping.resolves({});
+
       return indexCache.init()
         .then(() => {
           should(listIndexesStub.calledOnce).be.true();
@@ -49,6 +49,86 @@ describe('Test: core/indexCache', () => {
           should(indexCache.indexes.foo).be.an.Array().and.match(['bar', 'baz', 'qux']);
         });
     });
+
+    it('should inject the default mapping if none found', () => {
+      return indexCache.init()
+        .then(() => {
+          should(kuzzle.internalEngine.updateMapping)
+            .be.calledThrice();
+
+          should(indexCache.defaultMappings.foo)
+            .eql(indexCache.commonMapping);
+        });
+    });
+
+    it('should re-inject the existing mapping if one found', () => {
+      kuzzle.internalEngine.getFieldMapping
+        .resolves({
+          foo: {
+            mappings: {
+              baz: {
+                '_kuzzle_info.updater': {
+                  mapping: {
+                    some: 'mapping'
+                  }
+                }
+              }
+            }
+          }
+        });
+
+      return indexCache.init()
+        .then(() => {
+          should(indexCache.defaultMappings.foo).eql({
+            _kuzzle_info: {
+              properties: Object.assign({}, indexCache.commonMapping._kuzzle_info.properties, {some: 'mapping'})
+            }
+          });
+
+          for (const collection of ['bar', 'baz', 'qux']) {
+            should(kuzzle.internalEngine.updateMapping)
+              .be.calledWithMatch(collection, {
+                properties: {
+                  _kuzzle_info: {
+                    properties: {
+                      some: 'mapping'
+                    }
+                  }
+                }
+              });
+          }
+        });
+    });
+
+    it('should inject kuzzle_info mapping for known fields only', () => {
+      kuzzle.internalEngine.getFieldMapping
+        .resolves({
+          foo: {
+            mappings: {
+              baz: {
+                '_kuzzle_info.unknown': {
+                  mapping: {
+                    some: 'mapping'
+                  }
+                }
+              }
+            }
+          }
+        });
+
+      return indexCache.init()
+        .then(() => {
+          should(indexCache.defaultMappings.foo).eql(indexCache.commonMapping);
+
+          for (const collection of ['bar', 'baz', 'qux']) {
+            should(kuzzle.internalEngine.updateMapping)
+              .be.calledWithMatch(collection, {
+                properties: indexCache.commonMapping
+              });
+          }
+        });
+    });
+
   });
 
   describe('#initInternal', () => {

--- a/test/mocks/kuzzle.mock.js
+++ b/test/mocks/kuzzle.mock.js
@@ -107,6 +107,7 @@ class KuzzleMock extends Kuzzle {
 
     this.indexCache = {
       indexes: {},
+      defaultMappings: {},
       add: sinon.stub(),
       exists: sinon.stub(),
       init: sinon.stub().resolves(),
@@ -134,6 +135,8 @@ class KuzzleMock extends Kuzzle {
       exists: sinon.stub().resolves(),
       expire: sinon.stub().resolves(),
       get: sinon.stub().resolves(foo),
+      getFieldMapping: sinon.stub().resolves(),
+      getMapping: sinon.stub().resolves(),
       mget: sinon.stub().resolves({hits: [foo]}),
       index: 'internalIndex',
       init: sinon.stub().resolves(),
@@ -145,8 +148,7 @@ class KuzzleMock extends Kuzzle {
       scroll: sinon.stub().resolves(),
       search: sinon.stub().resolves(),
       update: sinon.stub().resolves(),
-      updateMapping: sinon.stub().resolves(foo),
-      getMapping: sinon.stub()
+      updateMapping: sinon.stub().resolves(foo)
     };
 
     this.once = sinon.stub();

--- a/test/services/implementations/elasticsearch.test.js
+++ b/test/services/implementations/elasticsearch.test.js
@@ -1360,6 +1360,37 @@ describe('Test: ElasticSearch service', () => {
           should(elasticsearch.config.commonMapping).eql(mapping);
         });
     });
+
+    it('should reuse a previously defined common mapping', () => {
+      kuzzle.indexCache.defaultMappings[index] = {
+        gordon: { type: 'text' }
+      };
+
+      return elasticsearch.updateMapping(new Request({
+        index,
+        collection,
+        body: {
+          properties: {
+            freeman: { type: 'boolean' }
+          }
+        }
+      }))
+        .then(() => {
+          const esReq = elasticsearch.client.indices.putMapping.firstCall.args[0];
+
+          should(esReq).eql({
+            index,
+            type: collection,
+            body: {
+              properties: {
+                gordon: { type: 'text' },
+                freeman: { type: 'boolean' }
+              }
+            }
+          });
+        });
+
+    });
   });
 
   describe('#getMapping', () => {
@@ -1599,6 +1630,44 @@ describe('Test: ElasticSearch service', () => {
         });
     });
 
+    it('should reuse a previously created common mapping', () => {
+      elasticsearch.client.indices.putMapping.resolves({});
+      elasticsearch.kuzzle.indexCache.exists.returns(true);
+
+      request.input.resource.collection = '%foobar';
+
+      kuzzle.indexCache.defaultMappings[index] = {
+        gordon: { type: 'text' }
+      };
+
+      const collectionMapping = {
+        properties: {
+          freeman:  { type: 'keyword' },
+          _kuzzle_info: {
+            properties: {
+              author: { type: 'keyword' }
+            }
+          }
+        }
+      };
+      request.input.body = collectionMapping;
+
+      return elasticsearch.createCollection(request)
+        .then(() => {
+          const esReq = elasticsearch.client.indices.putMapping.firstCall.args[0];
+
+          should(esReq.body['%foobar'].properties).eql({
+            gordon:   { type: 'text' },
+            freeman:  { type: 'keyword' },
+            _kuzzle_info: {
+              properties: {
+                author: { type: 'keyword' }
+              }
+            }
+          });
+        });
+
+    });
   });
 
   describe('#truncateCollection', () => {

